Add support for a Recycler-managed "Host Heap" to facilitate tracing of objects which relate to scriptable types

BoCupp opened the pull request:
This change exposes functions that will allocate memory inside of Chakra's GC. This memory can be precisely traced or leaf, and either finalizable or not. Traced objects will have IRecyclerVisitedObject::Trace called on them during ProcessMark if they were added to a new type of mark stack during Mark. Trace implementations will use a new API to tell the marking context to add a set of pointers to the appropriate mark stack.

Finalizable objects will have Dispose called when the the object is no longer reachable, in order to perform unmanaged resource cleanup.

These objects (except for non-finalizable leaf objects) will all live inside a new heap block type. Currently there is only support for small and medium blocks.

GCStress has been updated to allocate objects of the various new types. Finalizable objects hold onto some unmanaged heap resources which would point out (via OOM) if objects are not correctly being cleaned up.

More details:
- Add support for small and medium RecyclerVisitedHostHeapBlockTypes
  - For now, add failfast for RecyclerVisitedObjects that are placed in the LargeHeapBlock
- Publish IRecyclerVisitedObject interface which adds new method (compared to FinalizableObject) for precise tracing support
  - The old FinalizableObject is now a specialization of RecyclerVisitedObject and is used internally by Chakra
  - For host allocations, the host is expected to provide an appropriate implementation of IRecyclerVisitedObject
- Publish IRecyclerHeapMarkingContext (used as the argument to IRecyclerVisitedObject::Trace)
  - The host must implement precisely tracing the (non-leaf) memory it allocates by calling IRecyclerHeapMarkingContext::MarkObjects from inside its implementation of IRecyclerVisitedObject::Trace
  - It is implemented by a stack wrapper on MarkContext that captures the parallel and interior template parameters from MarkContext::ProcessMark. We always use false for doSpecialMark, since all calls to Mark that originate from ProcessMark would pass false (the only place we pass true is when scanning the stack)
- Create Chakra exports for allocation and rootaddref/release functions
- Introduce a new RecyclerHeap lib in chakra that defines the exports in the cpp and declares them in the h
- Add support for RecyclerVisitedObjects in GCStress

Note: this change is derived from a rebase of work done by Daniel Libby - all credit to him for the work to enable this concept.

Penguinwizzard commented:
Could you rebase this PR on top of master? We have fixes for the CI issues (some of which you addressed in your own changes) in the current version of master.

digitalinfinity commented on lib/Common/Memory/HeapBlock.h:
```diff
@@ -221,31 +257,58 @@ class HeapBlock
-        MediumAllocBlockTypeCount = 5, // Actual number of types for blocks containing medium allocations
-        SmallBlockTypeCount = 11,      // Distinct block types independent of allocation size using SmallHeapBlockT
+#ifdef RECYCLER_VISITED_HOST
+        SmallAllocBlockTypeCount = 7, // Actual number of types for blocks containing small allocations
```
There's no test for this today IIRC but you probably need to update JD to handle this update?

BoCupp replied:
Yes, I believe it has been updated, but that code lives outside of ChakraCore and isn't part of this PR, or is there a spot under core that needs to be updated?

digitalinfinity commented on lib/Common/Memory/HeapBlockMap.inl:
```diff
@@ -142,13 +142,38 @@ HeapBlockMap32::Mark(void * candidate, MarkContext * markContext)
 #endif
         ((SmallFinalizableHeapBlock*)chunk->map[id2])->ProcessMarkedObject<doSpecialMark>(candidate, markContext);
         break;
+#ifdef RECYCLER_VISITED_HOST
```
For visited objects, is mark always called with interior marking mode- if so, do we actually need to implement this or should we just assert/fail fast here?

BoCupp replied:
We do need to implement it.  This is the core path.  The interior template parameter of MarkingContext::Mark controls if we are using HeapBlockMap*::MarkInterior or HeapBlockMap*::Mark.  The interior template parameter though I believe is only true for the MemGC recycler, so the JS GC is always using HeapBlockMap*::Mark.

Inside HeapBlockMap*::Mark we must still support having interior pointers to the new types of blocks we added, since we expect the majority of our pointers to be interior due to our wrapper approach for implementing IRecyclerVisitedObject in the host.  As a result, once we know the block type in the HeapBlockMap*::Mark method, we can support interior references by calling MarkInteriorInternal on a computed pointer obtained from GetRealAddressFromInterior.

BoCupp replied:
I think I could assert/fail in the markinterior version.  I'll try adding that.

BoCupp replied:
The other place which sets enableScanInteriorPointers (which leads to calling HeapBlockMap*::MarkInterior) is in the Recycler constructor based on a CUSTOM_CONFIG_FLAG.  If I force failure under the MarkInterior path then allocating our new type of heap block would lead to a failure in that configuration.  Is that OK or should MarkInterior stay as is?

digitalinfinity commented on lib/Common/Memory/HeapBucket.cpp:
```diff
@@ -1617,6 +1623,9 @@ HeapBucketGroup<TBlockAttributes>::ScanInitialImplicitRoots(Recycler * recycler)
 #endif
     finalizableHeapBucket.ScanInitialImplicitRoots(recycler);
+#ifdef RECYCLER_VISITED_HOST
+    recyclerVisitedHostHeapBucket.ScanInitialImplicitRoots(recycler);
```
Is this needed? Can visited host objects ever be implicit roots?

BoCupp replied:
I believe only the MemGC allocators set that bit, so this bucket should never contain any "implicit roots".  I'll try removing it.

digitalinfinity commented on lib/Common/Memory/HeapBucket.cpp:
```diff
@@ -1651,6 +1663,9 @@ void
 {
     finalizableHeapBucket.Sweep(recyclerSweep);
+#ifdef RECYCLER_VISITED_HOST
+    recyclerVisitedHostHeapBucket.Sweep(recyclerSweep);
```
Curious- is there a reason that these need to be swept before the smallFinalizableWithBarrier heap bucket? Or was this just an arbitrary choice?

BoCupp replied:
I believe its arbitrary.  I think it was because Daniel first added the code without the RECYCLER_VISITED_HOST directives and it looked neater organizing the non-conditionally-compiled code together.

digitalinfinity commented on lib/Common/Memory/MarkContext.h:
```diff
@@ -102,6 +106,7 @@ class MarkContext
     Recycler * recycler;
     PagePool * pagePool;
     PageStack<MarkCandidate> markStack;
+    PageStack<IRecyclerVisitedObject*> preciseStack;
```
Consider wrapping this in #ifdef RECYCLER_VISITED_HOST?

BoCupp replied:
Yes, OK.

digitalinfinity commented on lib/Common/Memory/Recycler.inl:
```diff
@@ -517,6 +521,13 @@ Recycler::AddMark(void * candidate, size_t byteCount) throw()
+Recycler::AddPreciselyTracedMark(IRecyclerVisitedObject * candidate) throw()
+{
+    // This is never called during parallel marking
+    Assert(this->collectionState != CollectionStateParallelMark);
```
Is this ever called during mark or only during rescan? Can we turn this assert into the positive case instead of the negative case?

BoCupp replied:
Only during rescan.  I can change the assert.

BoCupp replied:
Actually, when changing the assert and running GCStress I find that while called under BackgroundRescan the Recycler's collectionState is Collection_ConcurrentMark... which doesn't have the Collection_Rescan flag.  I think the important aspect of the assert, however, is captured in that a Recycler API which pushes something into a MarkingContext lacks the information necessary to know which MarkingContext in the case of parallel marking.

I made a minor change to the assert to make this clearer:

    // This API cannot be used for parallel marking as we don't have enough information to determine which MarkingContext to use.
    Assert((this->collectionState & Collection_Parallel) == 0);

digitalinfinity commented on lib/Common/Memory/Recycler.inl:
```diff
@@ -36,18 +36,22 @@ namespace Memory
 Recycler::AllocWithAttributesInlined(DECLSPEC_GUARD_OVERFLOW size_t size)
 {
-    // All tracked objects are client tracked objects
+    // All tracked objects are client tracked or recycler host visited objects
```
Confused by this comment- why are host visited objects allocated with the track bit?

BoCupp replied:
We are currently using the TrackBit to indicate an object that needs tracing.  All our RecyclerVisitedHostTraced* objects currently have the TrackBit set.  Seems like we could have chosen to use "not Leaf" for the same purpose.  I'll explore that option.

digitalinfinity commented on lib/Common/Memory/SmallFinalizableHeapBlock.cpp:
```diff
@@ -103,6 +141,146 @@ SmallFinalizableHeapBlockT<TBlockAttributes>::SetAttributes(void * address, unsi
+{
+    bool noOOMDuringMark = true;
+
+    if (attributes & TrackBit)
```
I don't see any allocation variant in Recycler.h where we can allocated visited host objects with a track bit so when would this condition be true? Wondering if the track/newtrack code can be just removed from this method and treated as assert/failfast?

BoCupp replied:
See comment above on traced objects currently using the TrackBit.

digitalinfinity commented on lib/Common/Memory/SmallFinalizableHeapBlock.h:
```diff
@@ -134,6 +140,49 @@ class SmallFinalizableHeapBlockT : public SmallNormalHeapBlockT<TBlockAttributes
+    typedef TBlockAttributes HeapBlockAttributes;
+
+    static const ObjectInfoBits RequiredAttributes = RecyclerVisitedHostBit;
+    static const bool IsLeafOnly = true;
```
Why is this true? Also, is this used anywhere...?

digitalinfinity commented on lib/Common/Memory/SmallFinalizableHeapBlock.cpp:
```diff
@@ -103,6 +141,146 @@ SmallFinalizableHeapBlockT<TBlockAttributes>::SetAttributes(void * address, unsi
+        Assert((attributes & LeafBit) == 0);
+        Assert(block->GetAddressIndex(objectAddress) != SmallHeapBlockT<TBlockAttributes>::InvalidAddressBit);
+
+        if (!recycler->AddPreciselyTracedMark(reinterpret_cast<IRecyclerVisitedObject*>(objectAddress)))
```
Why is this not always called when we rescan a visited object?

BoCupp replied:
Objects are only pushed onto the preciseStack of the MarkContext if they need traced.  Since the TrackBit currently indicates whether or not we need traced it should guard AddPreciselyTracedMark.

BoCupp commented:
Note that I've addressed comments above not having to do with the use of the TrackBit for indicating which RecyclerVisitedObjects need the Trace method called.  I'm attempting to switch that to RecyclerVisitedObjects that are not leaves are traced, but GCStress is crashing.  When I resolve the issue I'll push again.

BoCupp commented:
Note that I've addressed comments above not having to do with the use of the TrackBit for indicating which RecyclerVisitedObjects need the Trace method called. I'm attempting to switch that to RecyclerVisitedObjects that are not leaves are traced, but GCStress is crashing. When I resolve the issue I'll push again.

digitalinfinity commented on lib/Common/Memory/SmallFinalizableHeapBlock.cpp:
```diff
@@ -103,6 +141,146 @@ SmallFinalizableHeapBlockT<TBlockAttributes>::SetAttributes(void * address, unsi
+template <bool doSpecialMark, typename Fn>
+bool SmallRecyclerVisitedHostHeapBlockT<TBlockAttributes>::UpdateAttributesOfMarkedObjects(MarkContext * markContext, void * objectAddress, size_t objectSize, unsigned char attributes, Fn fn)
+{
+    bool noOOMDuringMark = true;
```
Ok, so if I understand this correctly, HeapBlock::UpdateAttributesOfMarkedObjects should never be called for HostVisited Heap Blocks, correct? If so, in HeapBlock.inl, in that method, can you add an assert checking that?

BoCupp replied:
Yes that's right.  I'll add it.

BoCupp replied:
I pushed the change.

digitalinfinity commented on lib/Common/Memory/Recycler.inl:
```diff
@@ -36,18 +36,22 @@ namespace Memory
 class DummyVTableObject : public FinalizableObject
```
I think this should probably match the RecyclerVisitedObject vtable layout?

BoCupp replied:
FinalizableObject derives from IRecyclerVisitedObject so it has a matching v-table layout.

dilijev commented:
@dotnet-bot test Ubuntu static_ubuntu_linux_test please

dilijev commented:
@mmitche Sometimes we get stuck checks that completed but for some reason failed to report their completed status back to GitHub. Is there a way to force re-sending that status update instead of running the check from scratch?

mmitche commented:
@dilijev Nope :/

dilijev commented:
Oh well, small problem.
